### PR TITLE
[Impeller] Add render target builder.

### PR DESCRIPTION
**Pull request by guoguo338**
Fixed the issue https://github.com/flutter/flutter/issues/104006

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/entity/entity_pass.cc [hunk 1]
```diff
@@ -153,33 +154,33 @@ static RenderTarget CreateRenderTarget(ContentContext& renderer,
...
-        StorageMode::kDeviceTransient,     // color_storage_mode
-        StorageMode::kDevicePrivate,       // color_resolve_storage_mode
-        LoadAction::kDontCare,             // color_load_action
-        StoreAction::kMultisampleResolve,  // color_store_action
-        readable ? StorageMode::kDevicePrivate
-                 : StorageMode::kDeviceTransient,  // stencil_storage_mode
-        LoadAction::kDontCare,                     // stencil_load_action
-        StoreAction::kDontCare                     // stencil_store_action
-    );
+    return RenderTargetBuilder()
```
Comment (chinmaygarde) [inline impeller/entity/entity_pass.cc:157]: You can hoist the builder out of the conditional and set the common options like label, size, etc. and only set the MSAA specific builder options conditionally. That way, you don't have to repeat code in two places. Right now, this is a mechanical replacement of the existing calls.

File: impeller/entity/contents/content_context.cc [hunk 2]
```diff
@@ -249,9 +250,15 @@ std::shared_ptr<Texture> ContentContext::MakeSubpass(
 
   RenderTarget subpass_target;
   if (context->SupportsOffscreenMSAA()) {
-    subpass_target = RenderTarget::CreateOffscreenMSAA(*context, texture_size);
+    subpass_target = RenderTargetBuilder()
+                         .setSize(texture_size)
```
Comment (chinmaygarde) [inline impeller/entity/contents/content_context.cc:254]: See the comment about hoisting the builder out of the conditional elsewhere.

File: impeller/renderer/render_target_builder.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  stencil0.texture->SetLabel(SPrintF("%s Stencil Texture", label_.c_str()));
+
+  RenderTarget target;
+  target.SetColorAttachment(std::move(color0), 0u);
+  target.SetStencilAttachment(std::move(stencil0));
+
+  return target;
+}
+
+}  // namespace impeller
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.cc:157]: Nit: Newline at EOF.

File: impeller/renderer/render_target_builder.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#include "impeller/renderer/allocator.h"
+#include "impeller/renderer/formats.h"
+#include "render_target.h"
+
+namespace impeller {
+
+enum class RenderTargetType { Offscreen, OffscreenMSAA, kUnknown };
+
+class RenderTargetBuilder {
+ private:
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:22]: Nit: Private vars go at the bottom.

File: impeller/renderer/render_target_builder.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+#include "context.h"
+#include "flutter/fml/macros.h"
+#include "impeller/geometry/size.h"
+#include "impeller/renderer/allocator.h"
+#include "impeller/renderer/formats.h"
+#include "render_target.h"
+
+namespace impeller {
+
+enum class RenderTargetType { Offscreen, OffscreenMSAA, kUnknown };
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:19]: Nit: Follow the style guide with the `k` prefix for enum items.

File: impeller/renderer/render_target_builder.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+class RenderTargetBuilder {
+ private:
+  ISize size_;
+  std::string label_;
+
+  StorageMode color_storage_mode_;
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:26]: Default all ivars in the header with sensible defaults. Perhaps a device-transient texture.

File: impeller/renderer/render_target_builder.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    color_resolve_storage_mode_ = mode;
+    return *this;
+  }
+
+  RenderTargetBuilder& setRenderTargetType(RenderTargetType type) {
+    render_target_type_ = type;
+    return *this;
+  }
+
+  RenderTarget build(const Context& context);
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:97]: Perhaps a const function so then same builder can be used multiple times to create many render targets.

File: impeller/renderer/render_target_builder.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+}  // namespace impeller
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:100]: Nit: Newline at EOF.

File: impeller/renderer/render_target_builder.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    return *this;
+  }
+
+  RenderTargetBuilder& setRenderTargetType(RenderTargetType type) {
+    render_target_type_ = type;
+    return *this;
+  }
+
+  RenderTarget build(const Context& context);
+};
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:98]: Nit: FML_DISALLOW_COPY_AND_ASSIGN.

File: impeller/renderer/render_target_builder.h [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  RenderTarget CreateOffscreen(const Context& context);
+
+  RenderTarget CreateOffscreenMSAA(const Context& context);
+
+ public:
+  RenderTargetBuilder();
+
+  ~RenderTargetBuilder();
+
+  RenderTargetBuilder& setSize(ISize size) {
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:47]: Nit: OOL these and update the case to be CapitalCamelCase.

File: impeller/renderer/render_target_builder.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+enum class RenderTargetType { kOffscreen, kOffscreenMSAA, kUnknown };
+
+class RenderTargetBuilder {
+ public:
+  RenderTargetBuilder();
+
+  ~RenderTargetBuilder();
+
+  RenderTargetBuilder& SetSize(ISize size) {
```
Comment (chinmaygarde) [inline impeller/renderer/render_target_builder.h:27]: Can you OOL these too please? Put these in the implementation .cc file I mean.

**Review by chinmaygarde — APPROVED** [review]
Minor nit about the out of lining of the methods. But otherwise LGTM after the licenses are fixed. Thanks!

**Review by dnfield — APPROVED** [review]

File: impeller/aiks/picture.cc [hunk 12]
```diff
@@ -52,7 +53,10 @@ std::shared_ptr<Texture> Picture::RenderToTexture(
 
   // This texture isn't host visible, but we might want to add host visible
   // features to Image someday.
-  auto target = RenderTarget::CreateOffscreen(*context.GetContext(), size);
+  auto target = RenderTargetBuilder()
+                    .SetSize(size)
+                    .SetRenderTargetType(RenderTargetType::kOffscreen)
+                    .Build(*context.GetContext());
```
Comment (dnfield) [inline impeller/aiks/picture.cc:59]: This dropped the storage mode.

Comment (dnfield) [reply impeller/aiks/picture.cc:59]: (before this change the storage mode would end up as kDevicePrivate, now it ends up as kDeviceTransient)

Comment (dnfield) [reply impeller/aiks/picture.cc:59]: And a couple other defaults changed here too.

**Comment by dnfield** [comment]
I'm going to revert this change - it broke `Picture.toImage` and it's not entirely clear to me how to fix it forward. We really should have some kind of test for Picture.toImage somewhere to make sure this lands ok again.

File: impeller/aiks/picture.cc (hunk 12 above)
Comment (guoguo338) [reply impeller/aiks/picture.cc:59]: sorry to make wrong default values, now changed back.

**Pull request merged**